File: sdk/typescript/src/resolve.ts
import type { HttpClient } from "./http.ts";
import { ResolutionError } from "./errors.ts";

const UUID = /^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/i;

export interface Named {
  id: string;
  name?: string;
  [key: string]: unknown;
}

/**
 * Turns `{ agent: "reposage" }` into an id.
 *
 * Names, not ids, are the point: a script that reads `vault: "github-bot"` says
 * what it does, and a UUID in a README kills the demo. Resolution is exact
 * (case-insensitive) first, then a unique prefix, and a miss lists what the
 * account actually has — the error a caller can act on without opening the
 * console.
 */
export class Resolver {
  private readonly cache = new Map<string, Promise<Named[]>>();
  private readonly http: HttpClient;

  constructor(http: HttpClient) {
    this.http = http;
  }

  /** Drop the memoized listings — call after creating an agent elsewhere. */
  clear(): void {
    this.cache.clear();
  }

  list(path: string): Promise<Named[]> {
    let pending = this.cache.get(path);
    if (!pending) {
      pending = this.http.list<Named>(path).catch((error) => {
        // A failed listing must not be remembered as an empty account.
        this.cache.delete(path);
        throw error;
      });
      this.cache.set(path, pending);
    }
    return pending;
  }

  async resolve(path: string, what: string, nameOrId: string): Promise<Named> {
    const wanted = (nameOrId ?? "").trim();
    if (!wanted) throw new ResolutionError(`${what} is required (a name or id)`);

    // An id needs no listing — and an account with hundreds of agents should
    // not pay for one on every call.
    if (UUID.test(wanted)) {
      const known = (await this.peek(path))?.find((item) => item.id === wanted);
      return known ?? { id: wanted };
    }

    const items = await this.list(path);
    const byId = items.find((item) => item.id === wanted);
    if (byId) return byId;

    const lower = wanted.toLowerCase();
    const exact = items.filter((item) => (item.name ?? "").toLowerCase() === lower);
    if (exact.length === 1) return exact[0] as Named;
    if (exact.length > 1) {
      throw new ResolutionError(
        `More than one ${what} is named ${JSON.stringify(wanted)}. Use the id: ` +
          exact.map((item) => item.id).join(", "),
      );
    }

    const prefix = items.filter((item) => (item.name ?? "").toLowerCase().startsWith(lower));
    if (prefix.length === 1) return prefix[0] as Named;
    if (prefix.length > 1) {
      throw new ResolutionError(
        `${JSON.stringify(wanted)} matches more than one ${what}: ` +
          prefix.map((item) => item.name ?? item.id).join(", "),
      );
    }

    throw new ResolutionError(
      `No ${what} named ${JSON.stringify(wanted)}. On this account: ${describe(items)}`,
    );
  }

  async resolveId(path: string, what: string, nameOrId?: string | null): Promise<string | undefined> {
    if (!nameOrId) return undefined;
    return (await this.resolve(path, what, nameOrId)).id;
  }

  /** The cached listing if we already have one; never fetches. */
  private async peek(path: string): Promise<Named[] | null> {
    const pending = this.cache.get(path);
    if (!pending) return null;
    return pending.catch(() => null);
  }
}

function describe(items: Named[]): string {
  const names = items
    .map((item) => item.name ?? item.id)
    .filter(Boolean)
    .sort();
  return names.length ? names.join(", ") : "(none)";
}

File: sdk/python/src/fountain/resolve.py
"""Resolve human-friendly resource names to API ids."""

import re
import threading
from typing import Any, Dict, List, Optional

from .errors import ResolutionError
from .http import HttpClient

_UUID = re.compile(
    r"^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$", re.I
)


class Resolver:
    def __init__(self, http: HttpClient) -> None:
        self._http = http
        self._cache: Dict[str, List[Dict[str, Any]]] = {}
        self._lock = threading.Lock()

    def clear(self) -> None:
        with self._lock:
            self._cache.clear()

    def forget(self, path: str) -> None:
        with self._lock:
            self._cache.pop(path, None)

    def list(self, path: str) -> List[Dict[str, Any]]:
        with self._lock:
            cached = self._cache.get(path)
        if cached is not None:
            return cached
        items = self._http.list(path)
        with self._lock:
            self._cache[path] = items
        return items

    def resolve(self, path: str, what: str, name_or_id: str) -> Dict[str, Any]:
        wanted = (name_or_id or "").strip()
        if not wanted:
            raise ResolutionError("%s is required (a name or id)" % what)
        if _UUID.match(wanted):
            with self._lock:
                known = next(
                    (
                        item
                        for item in self._cache.get(path, [])
                        if item.get("id") == wanted
                    ),
                    None,
                )
            return known or {"id": wanted}

        items = self.list(path)
        by_id = next((item for item in items if item.get("id") == wanted), None)
        if by_id:
            return by_id
        lower = wanted.lower()
        exact = [item for item in items if str(item.get("name", "")).lower() == lower]
        if len(exact) == 1:
            return exact[0]
        if len(exact) > 1:
            raise ResolutionError(
                "More than one %s is named %r. Use the id: %s"
                % (what, wanted, ", ".join(str(item["id"]) for item in exact))
            )
        prefix = [
            item
            for item in items
            if str(item.get("name", "")).lower().startswith(lower)
        ]
        if len(prefix) == 1:
            return prefix[0]
        if len(prefix) > 1:
            raise ResolutionError(
                "%r matches more than one %s: %s"
                % (
                    wanted,
                    what,
                    ", ".join(str(item.get("name") or item["id"]) for item in prefix),
                )
            )
        names = sorted(str(item.get("name") or item.get("id")) for item in items)
        raise ResolutionError(
            "No %s named %r. On this account: %s"
            % (what, wanted, ", ".join(names) or "(none)")
        )

    def resolve_id(
        self, path: str, what: str, name_or_id: Optional[str]
    ) -> Optional[str]:
        return str(self.resolve(path, what, name_or_id)["id"]) if name_or_id else None
